doc/design/inference.md

```diff
@@ -15,91 +16,163 @@ predict = fluid.layers.fc(input=image,
+  // Initialize from buffer
+  ProgramBuilder(const char* buffer, const size_t num_bytes);
+
+  // Some utility interface maybe required by users
+  std::vector<std::string>& GetFeedVarNames() const;
+  std::vector<std::string>& GetFecthVarNames() const;
+  std::vector<int64_t> GetFeedVarShape(const size_t index);
+  std::vector<int64_t> GetFetchVarShape(const size_t index);
+
+  void AppendFetchVariables(const std::string& var_name);
+  ...
 
-Given an `inference_program`, it is easy to derive a `load_program` which is composed of `load_op` and is responsible for initializing all the parameter variables in `inference_program`. `load_program` will be executed once and `inference_program` can be executed as many times as the user need.
+  // Do transform to the inference program
+  ProgramBuilder* operator()(/* some optimizing strategy */);
+  ProgramBuilder* operator()(const std::vector<std::string>& feed_var_names,
+                             const std::vector<std::string>& fetch_var_names,
+                             /* some optimizing strategy */);
+
+ private:
+  framework::ProgramDesc* main_program_;
+  framework::ProgramDesc* startup_program_;
+  std::vector<std::string> feed_var_names_;
+  std::vector<std::string> fetch_var_names_;
+};
+```
 
-To summarize, an inferencer module should:
-- be initialized from files or from buffers
-- be composed of two `ProgramDesc`s, namely the `inference_program` and  `load_program`
+In the first design, `ProgramBuilder` contains all the elements memtioned above, and is instanced by protobuf message of the `main_program`. Other members `startup_program`, `feed_var_names` and `fetch_var_names` will also be derived in the constructor.
 
-All the initialization should be done in the constructor.
+There are two advantages of introducing an independent concept `ProgramBuilder`:
+- It is easy to add utility interfaces to support other requirements.
+  For example,
+  - `GetFeed/FetchVarNames`. It can be used to help users verify how many inputs and outputs there need and what the names are.
+  - `GetFeed/FetchVarShape`. It can be used to help users verify the size of each input and output.
+  - `AppendFetchVariables`. Normally, all fetching variables' names should be included in the protobuf message of the `main_program`. However, sometimes users may want to fetch extra variables for other use or debugging purposes, they can use this interface directly and have no need to regenerate the protobuf message again. Note that `main_program` may be modified in this interface.
+- It is possible to support online optimization of the inference program.
+  We will design an inference transpiler to do offline optimization for inference, which produce an optimized inference `ProgramDesc` for a given `ProgramDesc`. However, some optimization can be done online, such as
+  - changing the layout from `NCHW` to `NHWC`
+  - merging the computation of batch normalization layer to the front fc layer or conv layer
```

Comment:
@Xreki : Can you explain this merging of computation for batch norm (may not be needed for this doc)?

Reply:
Here I just list something maybe we can do in the future. 
About merging the computation of batch norm layer, you can find some detail [here](https://github.com/PaddlePaddle/Mobile/tree/develop/deployment/model/merge_batch_normalization)  and [here](https://github.com/PaddlePaddle/Paddle/pull/6704). After merging the batch norm layer, mobilenet can get a 30% speedup, without loss of precision.

Reply:
Oh nice. Thank you.